paper.md

```diff
@@ -318,19 +325,36 @@ The training conditions were exactly similar to the previous two case studies.
 
 As we can see from Figure III, Popper allowed us to run the workflow in a SLURM cluster with MPI and hence utilize the processing power of multiple GPUs and drastically reduce the training duration.
 
-# Results {#sec:result}
+### Setting up CI for our project
+
+We pushed our MNIST project to Github and activated the repository in Travis to set up continuous integration on our project.
+For long-running workflows like those consisting of ML/AI or BigData workloads, it is recommended to scale down various parameters like dataset size, epochs, etc. with the help of environment variables to reduce the CI running time and iterate quickly. 
+We declared environment variables like `NUM_EPOCHS`, `DATASET_REDUCTION`, and `BATCH_SIZE` to control the number of epochs, size of training data, and batch size respectively in our workflow.
+Using the above variables we used only 10% of the dataset and configured the training for a single epoch, thus effectively reducing our CI running time by approx. 75%.
+The `.travis.yml` file used by our case study is shown below. 
+It can be generated by running `popper ci travis` from the command line.
+
+```yaml
+---
+dist: xenial
+language: python
+python: 3.7
+services: docker
+install:
+- git clone https://github.com/systemslab/popper /tmp/popper
+- export PYTHONUNBUFFERED=1
+- pip install /tmp/popper/src
+script: popper run -f artifacts/local/.popper.yml -w artifacts/local
+```
 
-A summary of the training duration and accuracy obtained by running the workflow on three different computing environment is shown below.
+By setting up CI, users can continuously validate changes made to their workflows and also protect their workflows from getting outdated due to reasons such as outdated dependencies, outdated container images, broken links, etc.
 
-| Compute Environment    | Avg. Duration | Avg. Accuracy    |
-|------------------------|---------------|------------------|
-| Local Machine          |     4034.2s   |      99.16%      |
-| Kubernetes Cluster     |     1001.6s   |      99.236%     | 
-| Slurm Cluster          |     234.8s    |      98.69%      |
+# Results {#sec:result}
 
 ![Comparison of training durations in 3 different computing environments with Popper](./figures/plot.png){#fig:casestudies}
 
-From Table I & Figure III, it can be seen how the portability of Popper workflows drastically reduces software development and debugging time by enabling developers and researchers to quickly iterate and test in different computing environments.
+A summary of the training duration and accuracy obtained by running the workflow in three different computing environment is shown in Figure @Fig:casestudies.
+It can be seen from the graph how the portability of Popper workflows drastically reduces software development and debugging time by enabling developers and researchers to quickly iterate and test in different computing environments.
```

Comment:
We cannot claim that the graphs show proof of the portability, or minimal amount of effort for executing the workflow on distinct platforms.

If we want to show evidence that can somehow prove this, we can reference the configuration file differences between the three setups and show mention that it is "relatively" minimal (not absolute, as some readers might not agree that they are minimal).

the other thing we could show is the comparison of an associated README file (assuming it exists). So that readers can compare, a README in narrative form, which takes time to read, time to copy-paste bash commands from, and compare that agains the YAML workflow. This could be made in terms of lines of code, or show the README and YAML file side by side, in order to show that they're basically accomplishing the same, but for Popper we have a machine-readable, automated workflow that requires no human intervention